### PR TITLE
Release/1.5.0

### PR DESCRIPTION
**Pull request by ramirezfranciscof**
Setting up the branch and PR for the next release. There are still some more changes to incorporate I think, but at least we can start checking out the stuff and organizing the CHANGELOG (for example, how are we going to report all the changes in the docs? I don't think it is worth to put the details of all commits as it is now).

I couldn't manage to get the script `API_clients/github_pr_for_release.py` to work (is this python 2?), so I just ran `git log --pretty=oneline v1.4.2..` and made the changes on issue number manually.

### CODE REVIEW

**Comment by ramirezfranciscof** [comment]
1. What? Why did github copy here a conversation from another thread?
2. We would probably want to leave out of this release [the PR of the copy lists](https://github.com/aiidateam/aiida-core/pull/4516/commits/16bc30548f7f1c686d200935174533535e850fd5). How can I do this? I just rebase interactly from Master and drop that one?

**Comment by chrisjsewell** [comment]
> What? Why did github copy here a conversation from another thread?

I imagine its a conversation in an actual commit

> We would probably want to leave out of this release the PR of the copy lists.

why? if this PR breaks something, then I suggest it should be fixed before releasing v1.5.0

**Comment by chrisjsewell** [comment]
>  I don't think it is worth to put the details of all commits as it is now

No indeed, I'm very much against the changelog just being a list commits (you can use git or the github interface to see those); it should be a readable document.

For the documentation section for example, we should just explain the overall purpose of these commits is to progress/~complete the overall restructuring of our documentation (started in version ?)

and dropping python 3.5 support should be a headline

commits like "Pre-commit: reintroduce `pylint` rules" and "CI: add coverage patch threshold to prevent false positives" I don't think even need to be mentioned

**Comment by chrisjsewell** [comment]
(if you want to see my approach to changelogs 😉 https://github.com/executablebooks/MyST-Parser/blob/master/CHANGELOG.md, https://github.com/executablebooks/jupyter-book/blob/master/CHANGELOG.md)

**Comment by ramirezfranciscof** [comment]
> why? if this PR breaks something, then I suggest it should be fixed before releasing v1.5.0

I think the feature introduced is not that important right now (or at least I've seen no user request for this); it is mainly important for the Transfer calc I'm using and that will also eventually be used with the "remote archive" feature, but it is not crucial at the moment. Conversely, inadvertedly changing the behavior of the copy list may cause many plugins to break. Moreover, it is not clear what the fix should be since we have not a definite idea of how these list should behave (see [this issue](https://github.com/aiidateam/aiida-core/issues/4499)), I need to study/explore this a bit. For all of this, I would suggest not to rush this to the next release.

**Comment by sphuber** [comment]
I agree with most that @chrisjsewell has said:

1. The `CHANGELOG.md` needs to be improved.
    * Big changes, like dropping of Python version minor support needs to be announced at the top directly below the version number.
    * There is no need to list all the documentation commits. We can simply add a generic message that the documentation has been improved significantly (optionally listing some of the more impactful PRs). Note that I usually don't mention them because they are usually minor and don't really matter for a changelog. Exception was of course when we did the first documentation rehaul of v1.3.0 when it was mentioned.
   * A single "Other" section is not very useful. Please use the same headers that you see in preceding releases, i.e. "Bug fixes", "Improvements", "Changes", "Features" etc.
   * I think it is okay to keep some devops related changes in the log, but just move them in a "Devops" or "Developers" section toward the end
   * In general, I would try not to change the format too often. Having some consistency in this document would be nice.

2. It is possible to make a release branch without the commit that changed the behavior of the copy lists, but it will be a mess, because all the commits that follow after it will get new commit hashes and so ultimately, when we merge master back into develop after the release is merged, the commits will essentially be duplicated. I think I'd prefer to revert that commit in `develop` and simply do it properly with the issue you opened as a result of it.

**Comment by sphuber** [comment]
Also @ramirezfranciscof , did you run the helper scripts to add the headers to any new files? Check the contributors file to see if anyone needs to be added?

**Comment by chrisjsewell** [comment]
> In general, I would try not to change the format too often. Having some consistency in this document would be nice.

Definitely 👍 
We should add in https://github.com/aiidateam/aiida-core/wiki/How-to-make-a-new-release an outline of the changelog format and recommended section headings

**Comment by ramirezfranciscof** [comment]
No, the helper scripts returned `No files changed.`; did you detect some missing? I think I ran them as it was in the wiki.

Regarding the authors, I don't know what do you want to do: I think no one new joined since Chris and Marnik, and although a couple of persons there are no longer in the group, this was also the case when we released 1.4, so I wasn't sure the criteria for this and I didn't want to just take out people without discussing it.

And yes, the changelog need modifying, it is obviously not an intended viable version. As stated in the OP the idea was that I gathered all the issues and prepared the base format so that we would have everything in place to discuss the organization in the cathegories. I only separated the Docs since I was suggesting to "squash" that together, so it was easier to see what it implied. Anyways, ok, I can propose an initial organization with this feedback, I guess.

**Comment by sphuber** [comment]
> No, the helper scripts returned `No files changed.`; did you detect some missing? I think I ran them as it was in the wiki.

No, I didn't check it myself, I was just mentioning it to make sure you remembered. Fine if you ran the scripts.

> Regarding the authors, I don't know what do you want to do: I think no one new joined since Chris and Marnik, and although a couple of persons there are no longer in the group, this was also the case when we released 1.4, so I wasn't sure the criteria for this and I didn't want to just take out people without discussing it.

Same as above, I am not sure if anything needs to be changed. If you checked it and concluded there is no change needed, then that is ok.

**Comment by sphuber** [comment]
Two more comments @chrisjsewell and @ramirezfranciscof :

* Please go through the open PRs and analyse which still should or could go into the release and notify those involved of the release schedule.
* Some entries that are in the current change log have already been included in the patch releases v1.4. Make sure they are not included again.

**Comment by ltalirz** [comment]
Just fyi - @chrisjsewell has been working hard on the import/export; first refactoring the export and now the import https://github.com/aiidateam/aiida-core/pull/4510 to regain back some control over that part of the codebase and prepare the ground for archive export/import testing.
I will add importing of group extras in https://github.com/aiidateam/aiida-core/pull/4514 on top of chris's changes for importing and I would very much like to get this into 1.5.0. I'll be reviewing chris's PR today.

**Comment by ramirezfranciscof** [comment]
> Just fyi - @chrisjsewell has been working hard on the import/export; first refactoring the export and now the import #4510 to regain back some control over that part of the codebase and prepare the ground for archive export/import testing.
> I will add importing of group extras in #4514 on top of chris's changes for importing and I would very much like to get this into 1.5.0. I'll be reviewing chris's PR today.

Sure, no problem. Let me know when this is done so I'll rebase this branch and update the changelog.

> Some entries that are in the current change log have already been included in the patch releases v1.4. Make sure they are not included again.

I think I took those out now, thanks! This is why I originally mentioned using `git log --pretty=oneline v1.4.2..`; I should have maybe posed it directly as a question there, but I was wondering if this is correct or if there was a better way. I would have thought that would have shown me all different commits between the current branch and `v1.4.2`, but it also included some commits that were in both, not sure why (the ones you detected).

> Please go through the open PRs and analyse which still should or could go into the release and notify those involved of the release schedule.

There are a few PRs open regarding the documentation, but I don't think these would be ready for tomorrow. Anyways, since this would not be the end of the whole documentation tasks anyway, I don't personally think it is critical to have them for the release, but feel free to disagree.

> It is possible to make a release branch without the commit that changed the behavior of the copy lists, but it will be a mess, because all the commits that follow after it will get new commit hashes and so ultimately, when we merge master back into develop after the release is merged, the commits will essentially be duplicated. I think I'd prefer to revert that commit in develop and simply do it properly with the issue you opened as a result of it.

Yeah, I would be ok with reverting if that makes it easier than picking commits. Do I just need to [go to the PR](https://github.com/aiidateam/aiida-core/pull/4416) and click on the "Revert" button next to your merge?

**Comment by sphuber** [comment]
> I think I took those out now, thanks @sphuber ! This is why I originally mentioned using `git log --pretty=oneline v1.4.2..`; I should have maybe posed it directly as a question there, but I was wondering if this is correct or if there was a better way. I would have thought that would have shown me all different commits between the current branch and `v1.4.2`, but it also included some commits that were in both, not sure why (the ones you detected).

The reason is that those commits are in the history twice. The commits that went into 1.4.1 and 1.4.2 (or at least some of them) were committed directly to `develop`. However, this was done after Python 3.5 support had already been dropped from `develop`. Since `v1.4.*` is supposed to be Python 3.5 compatible, I had to make a release branch that branched off of before that point and cherry-pick the patch commits onto it. By doing so, they get a new hash, which were then merged into master, which then had to be merged back into develop, to make sure it is up to date with master. This final act ensures that now there are two commits on the `develop` branch that essentially represent the same change: one is the original, the other the cherry-pick. `git` can't distinguish this subtlety though and it sees them as two different commits, as if they would have been two randomly different commits. That is why those commits still show up in the command that you used, but really they have already been released. There is no magic bullet for this, which is why the release process, and especially the creation of the change log, remains a manual process that requires the person to be very aware of everything that happens in the codebase.

> There are a few PRs open regarding the documentation, but I don't think these would be ready for tomorrow. Anyways, since this would not be the end of the whole documentation tasks anyway, I don't personally think it is critical to have them for the release, but feel free to disagree.

I think this may have been mentioned during the group meeting, but I can't remember. Why is the release so pressing that it needs to be done tomorrow but these documentation PRs are not?

> Yeah, I would be ok with reverting if that makes it easier than picking commits. Do I just need to [go to the PR](https://github.com/aiidateam/aiida-core/pull/4416) and click on the "Revert" button next to your merge?

I don't know how to do it through the Github interface but on the command line it is easy.
```
git fetch --all
git checkout develop
git pull origin develop
git checkout -b revert/16bc30548f7f1c686d200935174533535e850fd5
git revert 16bc30548f7f1c686d200935174533535e850fd5
```
Then save the commit, push it and create a PR onto `develop`.

**Comment by ltalirz** [comment]
Thanks to great work by @chrisjsewell , import refactor and import of group extras is are now merged - yippie!

The commands already have `--verbosity` options, i.e. we might also want to get https://github.com/aiidateam/aiida-core/pull/3896 merged, but it's not absolutely necessary.

**Comment by ltalirz** [comment]
And there is one more thing we should decide how to handle group extras & group content upon import.
My proposal in https://github.com/aiidateam/aiida-core/issues/4383 is to have group extras react to the same flag as node extras (step 1).

In a second step, I propose to simplify the flags (merge `extras-mode-existing` and `extras-mode-new` in to `extras-mode`) and add a new flag `group-mode` with the same possible flag values, which determines how group contents (=which nodes they contain) is merged.

**Comment by ramirezfranciscof** [comment]
I have now updated with the commits from last rebase, feel free to add your comment @sphuber

File: CHANGELOG.md
Comment (sphuber) [inline CHANGELOG.md:6]:
```suggestion
In this minor version release, the support for Python 3.5 is dropped [[#4386]](https://github.com/aiidateam/aiida-core/pull/4386).
```

File: CHANGELOG.md
Comment (sphuber) [inline CHANGELOG.md:7]: Python 3.9 is also already out, but we don't support it yet
```suggestion
This version remains compatible with the following current Python versions that are not end-of-life:
```

File: CHANGELOG.md
Comment (sphuber) [inline CHANGELOG.md:18]:
```suggestion
- Refactor the archive reading code from the import implementation [[#4510]](https://github.com/aiidateam/aiida-core/pull/4510)
```

File: CHANGELOG.md
Comment (sphuber) [inline CHANGELOG.md:19]:
```suggestion
- Added remaining original documentation that didn't make it into the first step of the recent major overhaul of v1.3.0
```

File: CHANGELOG.md
Comment (sphuber) [inline CHANGELOG.md:28]:
```suggestion
- `CalcJob`: only attempt to clean up the retrieve temporary folder after parsing if it is present [[#4379]](https://github.com/aiidateam/aiida-core/pull/4379)
```

File: CHANGELOG.md
Comment (sphuber) [inline CHANGELOG.md:29]: Would make this an improvement

**Review by sphuber — CHANGES_REQUESTED** [review]
Thanks @ramirezfranciscof just a few suggestions for the changelog. All the changes so far in develop seem to be in, so list seems to be complete now

File: CHANGELOG.md
Comment (ramirezfranciscof) [reply CHANGELOG.md:29]: Just to make sure: I see selected all of the entries of bug fixes, you meant all of them? In that case I will directly eliminate the category.

Comment (sphuber) [reply CHANGELOG.md:29]: I was talking just about "Improve the deprecation....". We should keep the bug fix category

File: CHANGELOG.md
Comment (ramirezfranciscof) [reply CHANGELOG.md]: I changed this based on this recommendation, but I thought it would be better to adapt it to the format of the rest of the archive restructure logs. See first line:

```
- Archive importer: refactor the reading code [[#4510]](https://github.com/aiidateam/aiida-core/pull/4510)
- Archive importer: refactor the migrations code [[#4532]](https://github.com/aiidateam/aiida-core/pull/4532)
- Archive exporter: refactor progress bar [[#4504]](https://github.com/aiidateam/aiida-core/pull/4504)
- Archive exporter: decouple the extraction of data from the writing to file [[#4448]](https://github.com/aiidateam/
```

File: CHANGELOG.md
Comment (sphuber) [inline CHANGELOG.md:6]:
```suggestion
In this minor version release, support for Python 3.9 is added [[#4301]](https://github.com/aiidateam/aiida-core/pull/4301), while support for Python 3.5 is dropped [[#4386]](https://github.com/aiidateam/aiida-core/pull/4386).
```

File: CHANGELOG.md
Comment (sphuber) [inline CHANGELOG.md:14]:
```suggestion
- Process functions (`calcfunction` and `workfunction`) can now be submitted to the daemon just like `CalcJob`s and `WorkChain`s [[#4539]](https://github.com/aiidateam/aiida-core/pull/4539)
```

File: CHANGELOG.md
Comment (sphuber) [inline CHANGELOG.md:24]: Since these concern mostly refactoring and so have merely improvements for developers, should they go to the `Developers` section instead?

**Review by sphuber — CHANGES_REQUESTED** [review]

File: CHANGELOG.md
Comment (ramirezfranciscof) [reply CHANGELOG.md:24]: Yes, I was debating myself this when I was adding the new ones. I believe at the start we included the first of these here because they also had some performance improvements, but I agree it might just be better to send all of them to `Developers`.

File: CHANGELOG.md
Comment (chrisjsewell) [inline CHANGELOG.md:39]:
```suggestion
- Archive (import/export) refactor:
     * The refactoring goal was to pave the way for the implementation of a new archive format in v2.0.0 ([ aiidateam/AEP005](https://github.com/aiidateam/AEP/pull/21))
     * Three abstract+concrete interface classes are defined; writer, reader, migrator, which are **independent of the internal structure of the archive**. These classes are used within the export/import code.
     * The code in `aiida/tools/importexport` has been largely re-written, in particular adding `aiida/tools/importexport/archive`, which contains this code for interfacing with an archive, and **does not require connection to an AiiDA profile**.
     * The export logic has been re-written; to minimise required queries (faster), and to allow for "streaming" data into the writer (minimise RAM requirement with new format). It is intended that a similiar PR will be made for the import code.
     * A general progress bar implementation is now available in `aiida/common/progress_reporter.py`. All corresponding CLI commands now also have `--verbosity` option.
     * Merged PRs:
         * Refactor export archive ([#4448](https://github.com/aiidateam/aiida-core/pull/4448) & [#4534](https://github.com/aiidateam/aiida-core/pull/4534))
         * Refactor import archive ([#4510](https://github.com/aiidateam/aiida-core/pull/4510))
         * Refactor migrate archive ([#4532](https://github.com/aiidateam/aiida-core/pull/4532))
         * Add group extras to archive ([#4521](https://github.com/aiidateam/aiida-core/pull/4521))
         * Refactor cmdline progress bar ([#4504](https://github.com/aiidateam/aiida-core/pull/4504) & [#4522](https://github.com/aiidateam/aiida-core/pull/4522))
       * Deprecations: `export_zip`, `export_tar`, `export_tree`, `extract_zip`, `extract_tar` and `extract_tree` functions. `silent` key-word in the `export` function
       * Removed: `ZipFolder` class
```

**Review by chrisjsewell — CHANGES_REQUESTED** [review]
Basically copied from the update in the last aiida team meeting 😄

File: CHANGELOG.md
Comment (ramirezfranciscof) [reply CHANGELOG.md]: Since it is a pretty big change with a considerable explanation, I think it would be better to make it a separate section. Check out how it looks now and let me know what you think.

File: CHANGELOG.md
Comment (chrisjsewell) [inline CHANGELOG.md:47]: typo:

```suggestion
- The export logic has been re-written; to minimise required queries (faster), and to allow for "streaming" datainto the writer (minimise RAM requirement with new format). It is intended that a similiar PR will be made for the import code.
```

File: CHANGELOG.md
Comment (chrisjsewell) [reply CHANGELOG.md]: yep thats fine ta 👍

File: CHANGELOG.md
Comment (chrisjsewell) [inline CHANGELOG.md:39]:
```suggestion
- CI: add coverage patch threshold to prevent false positives [[#4413]](https://github.com/aiidateam/aiida-core/pull/4413)
- CI: Allow for mypy type checking of third-party imports [[#4553]](https://github.com/aiidateam/aiida-core/pull/4553)
```

File: CHANGELOG.md
Comment (chrisjsewell) [inline CHANGELOG.md:56]:
```suggestion
- Updated archive version from `0.9` -> `0.10`
- Deprecations: `export_zip`, `export_tar`, `export_tree`, `extract_zip`, `extract_tar` and `extract_tree`functions. `silent` key-word in the `export` function
```

**Review by chrisjsewell — CHANGES_REQUESTED** [review]
add final merged PRs

**Review by sphuber — APPROVED** [review]

**Review by chrisjsewell — APPROVED** [review]

**Pull request merged**
